[Jed] Step 3-4 : 사진 추가하기 기능 구현

junu0516 opened the pull request:
## 작업 내용
- [x] 사진 추가하기 버튼 추가
- [x] 사진 추가하기 기능 구현
- [x] 사각형 모델에 선택한 사진 바이너리 데이터 보관
- [x] 사진 데이터 관리하는 뷰, 모델 생성

## 고민과 해결
- 기존에 언급한 뷰컨트롤러상에서 현재 터치된 뷰를 저장하는 __`self.currentlyTouchedView`__ 를 제거했습니다.
    - subViews 배열 내 인덱스와 self.plane 내부의 rectangle 배열 내 인덱스를 서로 맞췄습니다.
- __`Rectangle.Alpha`__ 값을 초기화할 때 열거형의 allCases를 사용하던 것에서 탈피하고자 타입을 구조체로 바꾸고, 내부에 다시 열거형을 가진 형태로 수정했습니다.
    - 이후 아무 정수값이나 해당 구조체의 생성자 파라미터로 넘겨서 생성자 내부에서 값을 적절하게 조정해서 열거형의 케이스 중 적절한 것을 선택하도록 했습니다. 
    
- 사용자가 랜덤으로 선택한 색, 사용자가 선택한 사진을 각각 배경색으로 지정하는 __`ColorRectangle`__ , __`ImageRectangle`__ 클래스를 구분하여 생성했습니다.
    - 기존에 사용하던 __`Rectangle`__ 클래스가 두 클래스의 부모 클래스가 되도록 했습니다.
    - 기존의 __`Rectangle`__ 내부의 __`Rectangle.Color`__ 속성은 __`ColorRectangle`__ 내부로 이동시켰습니다.   
    
- 추후 기존에 언급하신 리팩토링 피드백 사항 중 아직 완전히 수정 못한 부분은 지속적으로 보완해나가려고 합니다!
    - RectangleFactory, RectangleViewFactory의 리턴값이 구체적인 타입에 의존하지 않는 부분에 대해 고민하고 있습니다.
    - 지금은 Rectangle을 ColorRectangle, ImageRectangle이 각각 상속받는 형태이지만 이를 지난 수업시간에 말씀해주신 것을 참고해서 프로토콜 지향으로 수정하고자 합니다.

junu0516 committed:
Refactor: Rectangle.Alpha 타입 구조체로 변경

- 외부에서 opacity를 정수값으로 받아서 열거형의 allCases 배열의 인덱스로 활용하던 것에서, 구조체의 생성자 파라미터로 받아 내부에서 다시 초기화하도록 수정
- Rectangle.Alpha 내부에 10가지 케이스를 가진 Opacity 열거형을 별도로 선언
- Alpha의 생성자 내부에서 Opacity의 케이스 중 적절한 것을 골라 초기화하는 것

junu0516 committed:
Refactor: Plane에서 Rectangle.Id값 받는 subscript 제거

- 기존의 selectedRectangleIndex 대신 아얘 Rectangle 옵셔널 타입의 selectedRectangle으로 대체
- Rectangle.Id 값 받는 subscript 모두 제거

junu0516 committed:
Revert "Refactor: Plane에서 Rectangle.Id값 받는 subscript 제거"

This reverts commit cf138a11a43f3f49350d712e123ca0ebd9ea5b35.

junu0516 committed:
Revert "Revert "Refactor: Plane에서 Rectangle.Id값 받는 subscript 제거""

This reverts commit 3e5a24bbd6564f2bc3232e93c99fbcec8420b231.

junu0516 committed:
Refactor: Plane 내부에서 제거했던 Rectangle 인덱스 속성 복구

- selectedRectangle 다시 제거
- selectedRectangleIndex 속성 복구
- selectedRectangleIndex 값을 subscript 내부에서 결정하도록 수정
- subscript 내부 반복문은 인덱스 기반으로 수행되도록 수정

junu0516 committed:
Refactor: currentlyTouchedView 속성 제거

- Plane에서 특정 선택된 Rectangle 모델 찾으면, 해당 모델의 인덱스값도 같이 notify하도록 수정
- ViewController에서는 전달받은 인덱스값으로 canvasView.subViews 배열에서 해당 인덱스에 위치한 RectangleView를 찾아 업데이트

junu0516 committed:
Refactor: RectangleView 서브뷰 클래스 제거

- 사실상 역할이 없기 때문에 클래스 정의 제거
- RectangleView의 구분은 typealias로 대체

junu0516 committed:
Feat: CanvasView에 사진 추가 버튼 생성

junu0516 committed:
Add: ColorRectangle, ImageRectangle 추가

- 이미지 배경, 사용자 정의 색 배경이 각각 다른 Rectangle 서브 클래스

junu0516 committed:
Refactor: Rectangle 클래스 원형 클래스 형태로 수정

- Rectangle.Color 타입은 ColorRectangle에서 사용하기 때문에 제거

junu0516 committed:
Refactor: RectangleFactory, RectangleViewFactory 수정

- 각각 ColorRectangle, ImageRectangle 관련 모델, 뷰를 생성하도록 메소드를 달리함

junu0516 committed:
Refactor: Plane 클래스 내부에서 배경색 바꿀 때 ColorRectangle 타입 사용하도록 수정

junu0516 committed:
Feat: 이미지 사각형 추가 기능 생성

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -1,13 +1,14 @@
 import UIKit
 import OSLog
 
+typealias ColorRectangleView = UIView
+typealias ImageRectangleView = UIImageView
```
특별히 하위 타입이 없으면 typealias도 좋습니다만 선언하려면 여기가 아니라 뷰 파일을 분리하는 게 좋습니다. 
그래야 ColorRectangleView나 ImageRectangleView 타입을 바꾸더라도 여기를 고치지 않고 해당 파일만 수정하면 되겠죠

junu0516 replied:
넵 typealias 선언 부분은 별도 파일로 분리하겠습니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -65,28 +65,54 @@ class ViewController: UIViewController{
     @objc func addingRectangleCompleted(_ notification: Notification) {
         guard let rectangle = notification.userInfo?[Plane.UserInfoKey.rectangleAdded] as? Rectangle else { return }
         guard let canvasView = self.canvasView else { return }
-        let rectangleView = RectangleViewFactory.createRectangleView(rectangle: rectangle)
-        canvasView.insertSubview(rectangleView, belowSubview: canvasView.generatingButton)
+        
+        if let rectangleView = createRectangleView(rectangle: rectangle){
+            canvasView.insertSubview(rectangleView, belowSubview: canvasView.generatingButton)
+        }
+        
+    }
+    
+    private func createRectangleView(rectangle: Rectangle)-> UIView?{
+        if rectangle is ColorRectangle{
+            return RectangleViewFactory.createColorRectangleView(rectangle: rectangle as! ColorRectangle)
+        }else if rectangle is ImageRectangle{
+            return RectangleViewFactory.createImageRectangleView(rectangle: rectangle as! ImageRectangle)
+        }
```
여기 addingRectangleCompleted() -> createRectangleView() 흐름을 볼 때 
addingRectangleCompleted가 하나로 합쳐져서 전달되기 때문에 결국 createRectangleView에서 if - else 또는 switch-case로 분기해서 처리하게 됩니다. 
서로 다른 타입을 처리하는 거라면 이벤트를 분리하는 게 어떨까요? 그러면 서로 영향을 주지 않고 분기문 없이 처리가 가능하지 않을까요? 
어딘가에서 분기문으로 처리할꺼라면 흐름을 나누는 게 낫지 않나 생각해보세요

junu0516 replied:
넵 해당 부분은 createRectangleView() 의 각 조건문들을 별도의 이벤트 처리로 분리해보도록 하겠습니다!

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -65,28 +65,54 @@ class ViewController: UIViewController{
         guard let canvasView = self.canvasView else { return }
-        guard let rectangleView = self.currentlyTouchedView else { return }
-
+        let rectangleView = canvasView.subviews[selectedRectangleIndex]
```
canvasView.subviews 배열 index와 selectedRectangleIndex 를 맞추면 관리하기 무척 어려워집니다. 개수가 항상 동일하지 않을 수 있어서 예외코드나 늘어나게 되고 버그를 유발하게 되죠 
뷰는 항상 인덱스가 동일하지 않을 수 있거든요. 화면에 보여주는 방식에 따라서 subview가 더 추가될 수도 있어서 그렇습니다
그래서 뷰 인스턴스 자체와 rectangle 자체를 매칭하는 게 필요할 수 있습니다.

junu0516 replied:
한 가지 질문이 있는데, 말씀하신대로면 뷰 인스턴스와 rectangle을 매칭하고자 할 때 뷰 인스턴스가 rectangle의 아이디와 같은 모델값의 일부를 내부에 속성으로 가지고 있는 식으로 뷰-모델 간에 __`모델 속성의 일부를 공유해야 하는 것`__ 같습니다. 
이전까지는 뷰 인스턴스의 경우에는 직접적인 색상, 투명도, 좌표값 등을 제외한 아이디 문자열과 같은 나머지 모델의 속성은 가급적 뷰컨트롤러에서만 다루고 뷰 인스턴스는 이를 다루지 않도록 하려고 selectedRectangleView 속성을 임시로 두거나 혹은 위와 같이 모델의 인덱스와 뷰의 인덱스를 맞추는 식으로 진행하였습니다.
혹시 이런 상황에서 모델의 속성을 뷰 인스턴스가 속성으로 일부 가지고 있는 것도 괜찮을까요?

godrm replied:
인스턴스 자체를 뷰 컨트롤러를 둘 다 알게 되기 때문에 내부 속성을 모르고도 인스턴스끼리 매칭하면 됩니다.
물론 이것도 이게 베스트라고 할 수는 없는데요, 예를 들어 Rectangle을 가지고 View를 만들게 되는 시점에 둘 다 알기 때문에 
두 개를 매칭하는 구조를 갖고 있으면 됩니다.

junu0516 replied:
혹시 그러면 key-value 형식으로 뷰컨트롤러 상에서 사각형 모델과 사각형 뷰를 서로 매칭시켜서 보관하는 것이 말씀하신 __`Rectangle을 가지고 View를 만드는 시점에 두 개를 매칭하는 구조를 갖고 있는`__ 것의 예시가 될까요?
우선 뷰컨트롤러상에서 별도로 모델과 뷰 간의 매칭되는 정보를 가진 별도의 구조를 보관하라고 말씀하신 것으로 이해해서 맞게 이해한건지 여쭤보려 합니다!

godrm replied:
네. 둘 다 알 수 있는 곳이 ViewController 뿐이죠. 뷰는 모델을 모르고, 모델은 뷰를 몰라야 하니까요

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -65,28 +65,54 @@ class ViewController: UIViewController{
+        guard let stylerView = self.stylerView else { return }
+        let rectangle = rectangle as! ColorRectangle
         let r = rectangle.backgroundColor.r
         let g = rectangle.backgroundColor.g
         let b = rectangle.backgroundColor.b
-        let opacity = rectangle.alpha.opacity
+        let opacity = rectangle.alpha.opacity.rawValue
         let hexString = "#\(String(Int(r*255), radix: 16))\(String(Int(g*255), radix: 16))\(String(Int(b*255), radix: 16))"
-        stylerView.updateRectangleInfo(r: r, g: g, b: b, opacity: opacity, hexString: hexString)
-        canvasView.updateSelectedRectangleView(subView: rectangleView)
-        
+        stylerView.updateColorRectangleInfo(r: r, g: g, b: b, opacity: opacity, hexString: hexString)
+    }
+    
+    private func updateImageRectangleInfo(rectangle: Rectangle){
+        guard let stylerView = self.stylerView else { return }
+        stylerView.updateImageRectangleInfo(opacity: rectangle.alpha.opacity.rawValue)
     }
```
stylerView를 다루는 코드가 늘어나면 이 부분만 따로 StyleViewController를 만들어야 하는가 고민해보세요

junu0516 replied:
넵 스텝 진행하면서 ViewController가 비대해지는 것 같아서 추후에는 각각의 뷰에 맞춰 컨트롤러를 분리하고자 합니다! 이 부분은 추후 리팩토링 하면서 시도해보겠습니다

godrm replied:
어느 정도가 커지는 걸까요? 그 기준을 정하는 것도 중요하거든요

junu0516 replied:
음.. 우선은 육안으로 봤을 때 단순히 코드가 많아지는 것 같아서 분리하려고 했습니다.
조금 더 고민해보자면 맨 처음 뷰컨트롤러의 __`self.view`__ 를 StylerView, CanvasView의 두 부분으로 나눈 후 각각의 뷰에 각자의 기능을 추가하는 식으로 구현했기 때문에 애초에 기능이 점점 확장될 것을 염두에 두고 처음부터 뷰컨트롤러를 2개로 나눠 사용했어야 될 것 같습니다.
나누는 기준에 대해서는 좀 더 고민해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -119,29 +145,53 @@ class ViewController: UIViewController{
+    private func createImageRectangle(pickedImage: UIImage){
+        guard let canvasView = self.canvasView else { return }
+        guard let imageData = pickedImage.pngData() else { return }
+        let rectangle = RectangleFactory.createImageRectangle(maxX: canvasView.bounds.maxX - 50,
+                                                          maxY: canvasView.bounds.maxY - 250,
+                                                          minWidth: 150,
+                                                          minHeight: 150,
+                                                          maxWidth: 180,
+                                                          maxHeight: 180,
+                                                          image: imageData)
+        plane.addRectangle(rectangle)
```
이 부분도 plane 에게 rectangle를 생성해서 넘겨줄 것인가, 아니면 이벤트만 넘겨주고 plane이 생성할 것인가 고민해보세요. 
여기가 상위 모듈이니까 하위 모듈이 더 많은 일을 하는게 좋습니다.

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -60,17 +63,46 @@ class CanvasView: UIView{
     
+    private func setPhotoButtonAction(){
+        let buttonAction:UIAction = UIAction(title:""){_ in
+            self.sendPickingImageRequest()
+        }
+        self.photoButton.addAction(buttonAction, for: .touchDown)
+    }
+    
     private func sendCreatingRectangleRequest(){
         if let delegate = self.delegate{
             delegate.creatingRectangleRequested()
         }
```
이 정도면 그냥 buttonAction 내부에 `delegate?.creatingRectangleRequested()` 정도도 깔끔할 것 같습니다. 
다른 측면으로는 아예 UIAction을 외부에서 선언해서 넣을 수 있도록 하는 방법도 있습니다.

godrm commented on DrawingApp/DrawingApp/View/CanvasView.swift:
```diff
@@ -5,13 +5,16 @@ class CanvasView: UIView{
     
     weak var delegate: CanvasViewDelegate?
     private (set) var generatingButton: UIButton = UIButton()
+    private (set) var photoButton: UIButton = UIButton()
     private (set) var selectedRectangleView: UIView?
     
     init(frame: CGRect, backGroundColor: UIColor) {
         super.init(frame: frame)
         self.backgroundColor = backGroundColor
         setGeneratingButton()
+        setPhotoButton()
         setGeneratingButtonAction()
+        setPhotoButtonAction()
     }
```
CanvasView가 생성 버튼도 갖고 있고 사각형 뷰들을 그리는 역할도 하는군요
화면을 영역별로 View를 구분하는 것도 고려해보세요. CanvasView가 너무 많은 역할을 하는게 아닌가 생각해보면 좋습니다

junu0516 replied:
버튼 및 다른 UI 요소를 가지는 뷰를 추후 별도로 생성하려고 합니다.

godrm commented on DrawingApp/DrawingApp/Model/ColorRectangle.swift:
```diff
@@ -0,0 +1,40 @@
+import Foundation
+
+class ColorRectangle: Rectangle{
+    
```
ColorRectangle로 분리한 시도는 좋습니다

godrm commented on DrawingApp/DrawingApp/Model/ColorRectangle.swift:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+    
+    var backgroundColor: Color
```
항상 속성은 private var 혹은 private(set) 정도를 고려하세요

junu0516 replied:
앗 접근제어를 깜빡했네요! 바로 수정하겠습니다.

godrm commented on DrawingApp/DrawingApp/Model/ImageRectangle.swift:
```diff
@@ -0,0 +1,15 @@
+     
+    var backgroundImage: Data
+    override var description: String{
+        return "(\(id)), \(size), \(point), \(backgroundImage.hashValue), \(alpha)"
```
backgroundImage 같은 2진 데이터는 출력이 별로 의미는 없습니다만 이미지가 들어갔는 정도 보는 거라면 가능하겠네요.

junu0516 replied:
넵 해시값이 있다는 것을 일단 이미지가 뭐라도 잘 들어간 것이라고 판단해서 출력값으로 해시값을 사용했습니다.

godrm reviewed (approved):
구조나 큰 흐름에서는 괜찮은 것 같습니다. 
CanvasView가 조금 역할이 많은 것 같기도 합니다. self.view에 추가해도 되는 것도 있는 것 같구요
소소한 피드백을 남겼으니 머지하고 넘어갈 지, 열어놓고 개선할 지 판단해서 댓글 남겨주세요

junu0516 commented:
> canvasView.subviews 배열 index와 selectedRectangleIndex 를 맞추면 관리하기 무척 어려워집니다. 개수가 항상 동일하지 않을 수 있어서 예외코드나 늘어나게 되고 버그를 유발하게 되죠
뷰는 항상 인덱스가 동일하지 않을 수 있거든요. 화면에 보여주는 방식에 따라서 subview가 더 추가될 수도 있어서 그렇습니다
그래서 뷰 인스턴스 자체와 rectangle 자체를 매칭하는 게 필요할 수 있습니다.

 이 피드백에 대해서 이해가지 않는 부분이 있어 질문 남겼습니다! 대댓글 확인 부탁드립니다.

우선 해당 질문한 부분 먼저 개선해보고, 나머지 피드백은 다음 단계에서 같이 진행해보겠습니다!

junu0516 committed:
Refactor: CanvasView 내부 delegate 호출 코드 단순화

- if-let 사용하지 않고 ? 사용으로 수정

junu0516 committed:
Refactor: ColorRectangle.backgroundColor 속성 접근제어를 private(set)으로 변경

junu0516 committed:
Refactor: ViewController 내부에 rectangleDictionary 속성 추가

- Rectangle.Id:UIView 를 key-value로 받는 딕셔너리 속성 추가
- selectedRectangleIndex와 SubView 인덱스를 맞추지 않고, 파라미터로 받는 rectangle의 id 속성에 맞는 UIView value를 받아서 쓰도록 수정

junu0516 committed:
Refactor: rectangle 모델 생성 책임을 Plane으로 이동

- 하위모듈이기 때문에 모델 생성 책임을 뷰컨트롤러가 아닌 Plane에서 수행하도록 함

junu0516 commented:
피드백 해주신 부분들 중 아래 우선 개선해야 할 것들 수정해서 push했습니다! 확인부탁드립니다.
- delegate 옵셔널 판별 시 if-let 대신 ? 사용으로 단순화
- ColorRectangle.backgroundColor 속성 private (set)으로 변경
- 뷰컨트롤러에서 Rectangle 모델 인스턴스와 SubView 매칭하기 위한 딕셔너리 속성 추가
- 모델 생성 요청시, 팩토리로부터 모델을 생성하도록 호출하는 책임을 뷰컨트롤러가 아닌 Plane으로 이동

나머지 수정 사항은 머지해주시면 다음단계 진행하면서 같이 해보겠습니다.

godrm commented on DrawingApp/DrawingApp/Controller/ViewController.swift:
```diff
@@ -9,6 +9,7 @@ class ViewController: UIViewController{
     private var canvasView: CanvasView?
     private var stylerView: StylerView?
     private var plane: Plane = Plane()
+    private var rectangleDictionary:[Rectangle.Id:UIView] = [:]
```
Rectangle.Id 보다 Rectangle 타입 자체를 비교해보세요
그럼 Id 라는 값도 아마 private으로 처리할 수 있을 겁니다. 카드를 비교하던 것처럼요

junu0516 replied:
넵 Hashable을 마침 Id에 채택해놔서 이걸 키로 썼는데, 다시 보니 인스턴스 자체를 키로 하는 것이 말씀하신대로 속성을 굳이 가져와 쓸 필요가 없어 더 나을 것 같습니다! 
다음 PR때 같이 수정해서 푸시하겠습니다
